File: golang_demo/kafka_consumer_demo.go
package main

import (
	"cubetran_udf_golang/metadata"
	"encoding/json"
	"fmt"

	"github.com/confluentinc/confluent-kafka-go/kafka"
	"github.com/linkedin/goavro/v2"
)

// extract data from kafka and decode
func runKafkaDemo() {
	// kafka consumer
	consumer, _ := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": "localhost:9093",
		"group.id":          "example_group",
		"auto.offset.reset": "earliest",
	})
	_ = consumer.Subscribe("test", nil)

	// Avro codec
	codec, _ := goavro.NewCodec(metadata.AvroSchema)
	for {
		msg, _ := consumer.ReadMessage(-1)
		apeDtsRecord := bytes_to_struct(codec, msg.Value)
		jsonData, _ := json.Marshal(apeDtsRecord)
		fmt.Printf("ape_dts record: %s\n", jsonData)
	}
}

File: rust_demo/src/kafka_consumer_demo.rs
use rdkafka::{
    consumer::{BaseConsumer, Consumer},
    ClientConfig, Message,
};
use serde_json::json;

use crate::metadata::avro_converter::AvroConverter;

pub fn run_kafka_demo() -> anyhow::Result<()> {
    // Kafka consumer configuration
    let consumer: BaseConsumer = ClientConfig::new()
        .set("bootstrap.servers", "localhost:9093")
        .set("group.id", "example_group")
        .set("auto.offset.reset", "earliest")
        .create()?;

    // Subscribe to the topic
    consumer.subscribe(&["test"])?;
    let avro_converter = AvroConverter::new();
    loop {
        // Poll for messages
        match consumer.poll(None) {
            Some(Ok(msg)) => {
                if let Some(payload) = msg.payload() {
                    let ape_dts_record = avro_converter.avro_to_ape_dts(payload)?;
                    println!("ape_dts record: {}", json!(ape_dts_record))
                }
            }
            Some(Err(e)) => {
                eprintln!("Error while receiving message: {:?}", e);
            }
            None => {
                // No message received, continue polling
            }
        }
    }
}
